### PR TITLE
Java Exception Handling

### LINKED ISSUE #406: [Java] Add exception handling
# Description

Exception handling for Java aren't presently on Code Thesaurus. This issue is to add that concept to the system.

# Directions

* On the home page, on the right side (the reference sheet side), choose the language and version, and then the concept.
* The page says the data doesn't exist, but clicking the arrow for the template will provide the template and file for you to use.
* Save that file in `/web/thesauruses/[language]/[version]/[concept].json`.
* Fill out all of the `"code"` sections.
* Add in `"not-implemented": "true"` if the concept doesn't exist in the language. You can then delete the `"code"` line.
* Add in `"comment":` if you wish to explain something in the code
* Delete the `"name":` fields before your PR
* Save, commit your code, push to your fork, then make a PR.
* Watch the checks to make sure they all pass. If they don't, click on "Details" and it should tell you what to modify.

# Claiming This Issue

Please leave a comment below so I can assign you to it. This is to prevent multiple people from working on the same issue. 

Check out [CONTRIBUTING.md](https://github.com/codethesaurus/codethesaur.us/blob/main/CONTRIBUTING.md) for more information on how to contribute. 

Check out [https://docs.codethesaur.us/thesaurus/](https://docs.codethesaur.us/thesaurus/) for information on how Code Thesaurus works and how to add to the thesauruses.

# Need Help?

Leave a comment on here with your question, or feel free to ask [@geekygirlsarah](https://twitter.com/geekygirlsarah) on Twitter. You can also email coreteam@codethesaur.us too.

# Thanks!

We really value and appreciate the additions people make to Code Thesaurus, no matter how large or small. So thank you for taking the time to work on this!

### PR DESCRIPTION
**Pull request by Shreya-7**
## What GitHub issue does this PR apply to?

 

Resolves #406

## What changed and why?

This code adds the exception handling reference for Java versions 11, 15 and 17.

## Checklist

   

- [x] I claimed any associated issue(s) and they are not someone else's
- [x] I have looked at documentation to ensure I made any revisions correctly
- [x] I tested my changes locally to ensure they work
- [ ] (If editing Django) I have added or edited any appropriate unit tests for my changes

## Any additional comments or things to be aware of while reviewing?

There are no differences in exception handling between the 3 versions that I could find, except for enhanced messages in NullPointerExceptions.

### CODE REVIEW

**Comment by Shreya-7** [comment]
@geekygirlsarah can you review this PR please? :)

File: web/thesauruses/java/11/exception_handling.json [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+            "name": "Statement to start code if no exceptions are caught",
+            "code": [
+                "bool success = true;",
+                "\ntry {",
+                "\n    something();",
+                "\n} catch (Exception e) {",
+                "\n    success = false;",
+                "\n}",
+                "\nif (success) {",
+                "\n}"
```
Comment (cafce25) [inline web/thesauruses/java/11/exception_handling.json]: ct will add extra newlines here resulting in double newline `"\n\n"` on the page, you can just remove \n when using an array as `code`

**Review by cafce25 — CHANGES_REQUESTED** [review]
Hey,
the issue with the double newlines occurs a couple of times, but apart from that this looks
very good!
Also adding multiple versions at once, nice!

File: web/thesauruses/java/11/exception_handling.json (hunk 1 above)
Comment (Shreya-7) [reply web/thesauruses/java/11/exception_handling.json]: Thanks for the callout, fixed!

**Comment by Shreya-7** [comment]
@cafce25 do we need something else here as well or are we good? :)

**Comment by cafce25** [comment]
I think it looks good, but only @geekygirlsarah can merge it.

File: web/thesauruses/java/11/exception_handling.json [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+                "bool success = true;",
+                "try {",
+                "    something();",
+                "} catch (Exception e) {",
+                "    success = false;",
+                "}",
+                "if (success) {",
+                "}"
+            ],
+            "comment": "Not natively possible, but we can use this hack to simulate Python's try-except-else"
```
Comment (geekygirlsarah) [inline web/thesauruses/java/11/exception_handling.json]: One of the "best practices" is that Code Thesaurus is a language reference tool and not an algorithm reference tool. So if you have to invent the functionality, it shouldn't be in CT. So this should just be a "not-implemented" line instead.

Comment (geekygirlsarah) [reply web/thesauruses/java/11/exception_handling.json]: (Same for similar places below you implement the functionality too)

File: web/thesauruses/java/11/exception_handling.json [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+                "}"
+            ]
+        },
+        "statement_to_end_trying_code": {
+            "name": "Statement to end a block of code to try ",
+            "not-implemented": true
+        },
+        "property_with_exception_code": {
+            "name": "Property within an exception object containing the error code",
+            "code": "cause"
```
Comment (geekygirlsarah) [inline web/thesauruses/java/11/exception_handling.json]: So this is more like `Exception.cause` instead of just `cause`?

**Review by geekygirlsarah — CHANGES_REQUESTED** [review]
For the moment I just glanced at the v11 file (not 15/17). Remove the invented functionality and check the exception properties (if those should be a class.function thing) and I'll come back to this again and hopefully merge it!

File: web/thesauruses/java/11/exception_handling.json (hunk 2 above)
Comment (Shreya-7) [reply web/thesauruses/java/11/exception_handling.json]: Makes sense, I've removed this part for all 3 versions. Haven't found any other such instances.

File: web/thesauruses/java/11/exception_handling.json (hunk 3 above)
Comment (Shreya-7) [reply web/thesauruses/java/11/exception_handling.json]: Hmm, more like `Exception.getCause()`. Nice catch, I've changed this part and added clarification.

**Comment by Shreya-7** [comment]
@geekygirlsarah do you mind reviewing this again? Would be great if it could get merged in time for Hacktoberfest :)

**Comment by geekygirlsarah** [comment]
> @geekygirlsarah do you mind reviewing this again? Would be great if it could get merged in time for Hacktoberfest :)

I get email notifications with every PR, issue comment, and push you do. I've gotten 7 email notifications from you alone. So I see the actions.

This is also a side project for me. Between a full time day job, mentoring high schoolers 2-3 times a week, on top of other life responsibilities, along with some really big and complicated PRs for this year's Hacktoberfest, it's taking me longer to review and merge them than I anticipated. I am getting to them all slowly and in the order they come in. So I appreciate your patience.

**Comment by Shreya-7** [comment]
Oh, wasn't aware, sorry about that 😶 Please take your time, no hurries!

**Comment by geekygirlsarah** [comment]
Thanks! I plan to get them all done by end of day Oct 31, but also I've merged 55 PRs so far and 20 more are open (11 needing changes or have failing tests). So there's a lot to juggle, but it's also my 4th Hacktoberfest with Code Thesaurus, so I make it work!

**Comment by Shreya-7** [comment]
That sounds so overwhelming, really appreciate it!

File: web/thesauruses/java/11/exception_handling.json (hunk 3 above)
Comment (geekygirlsarah) [reply web/thesauruses/java/11/exception_handling.json]: I was thinking you'd add the class in, but I guess saying it inherits from Throwable can work.

**Review by geekygirlsarah — APPROVED** [review]
Alright, I think it looks good and tests pass. I can merge it in.

And congrats on your first Code Thesaurus contribution! 🎉 

Thanks for adding this in, I appreciate it! And make sure you register for Hacktoberfest if you haven't, this project counts towards that.

**Pull request merged**